src/compute/src/compute_state/peek_stash.rs

```diff
@@ -0,0 +1,245 @@
+            peek.timestamp,
+            peek.literal_constraints.clone(),
+            oks_handle,
+        );
+
+        let task_handle = mz_ore::task::spawn(|| "compute::stash_peek_response", async move {
+            let start = Instant::now();
+
+            let result = Self::do_upload(
+                &persist_clients,
+                persist_location,
+                peek.uuid,
+                relation_desc,
+                rows_rx,
+            )
+            .await;
+
+            let result = match result {
+                Ok(peek_response) => peek_response,
+                Err(e) => PeekResponse::Error(e.to_string()),
+            };
+            match result_tx.send((result, start.elapsed())) {
+                Ok(()) => {}
+                Err((_result, elapsed)) => {
+                    debug!(duration = ?elapsed, "dropping result for cancelled peek {}", peek_uuid)
+                }
+            }
+        });
+
+        Self {
+            peek,
+            peek_iterator: Some(peek_iterator),
+            rows_tx: Some(rows_tx),
+            stashed_rows: None,
+            result: result_rx,
+            _abort_handle: task_handle.abort_on_drop(),
+        }
+    }
+
+    async fn do_upload(
+        persist_clients: &PersistClientCache,
+        persist_location: PersistLocation,
+        peek_uuid: Uuid,
+        relation_desc: RelationDesc,
+        mut rows_rx: tokio::sync::mpsc::Receiver<Result<Vec<(Row, NonZeroI64)>, String>>,
+    ) -> Result<PeekResponse, String> {
+        let client = persist_clients
+            .open(persist_location)
+            .await
+            .map_err(|e| e.to_string())?;
+
+        let shard_id = format!("s{}", peek_uuid);
+        let shard_id = ShardId::try_from(shard_id).expect("can parse");
+        let write_schemas: Schemas<SourceData, ()> = Schemas {
+            id: None,
+            key: Arc::new(relation_desc.clone()),
+            val: Arc::new(UnitSchema),
+        };
+
+        let result_ts = Timestamp::default();
+        let lower = Antichain::from_elem(result_ts);
+        let upper = Antichain::from_elem(result_ts.step_forward());
+
+        // We have to use SourceData, which is a wrapper around a Result<Row,
+        // DataflowError>, because the bare columnar Row encoder doesn't support
+        // encoding rows with zero columns.
+        //
+        // TODO: We _could_ work around the above by teaching the bare columnar
+        // Row encoder about zero-column rows.
+        let mut batch_builder = client
+            .batch_builder::<SourceData, (), Timestamp, i64>(shard_id, write_schemas, lower)
+            .await;
+
+        let mut num_rows: u64 = 0;
+
+        loop {
+            let row = rows_rx.recv().await;
+            match row {
+                Some(Ok(rows)) => {
+                    assert!(
+                        rows.len() > 0,
+                        "the contract is that we only get non-empty batches"
+                    );
```

Comment:
How important is this? It seems like we could just drop this assert and remove an assumption without any drawbacks?